script to generate addon layout

ocobles opened the pull request:
This script will help generate and define a new addon ensuring a standard structure for all them.

It mainly clones the https://github.com/equinix-labs/terraform-equinix-template template and modifies the files specific to this addons project. It also modifies the files of the main module to enable the new addon

Usage:

` ./generate_addon_structure.sh [ -n NAME ] [ -w DOCS_WEBSITE ]` 

(-h)       Show usage and brief help
(-n)       (Required) Name of the addon
(-w)       (Required) Addon official documentation home page URL

![image](https://user-images.githubusercontent.com/68897552/208113721-bc9de7c4-7cab-47b6-a9ba-f20ed7aea451.png)

Commit:
script to generate addon layout

Commit:
list modified/generated files and rename templates with extenstion

Commit:
improve script output and ensure newline EOF

Commit:
missing newline EOF

Commit:
fixup! missing newline EOF

Commit:
update usage description

displague commented on .utils/addon_specific_templates/examples/simple/variables.tf.tpl:
```diff
@@ -0,0 +1,36 @@
+# TEMPLATE: https://www.terraform.io/docs/language/values/variables.html
+# TEMPLATE: https://www.terraform.io/docs/language/expressions/types.html
+#
+variable "auth_token" {
```
We've found that in other projects "auth_token" and "project" can be too ambiguous. A `metal` prefix can help (and even that has been ambiguous in projects where metal means something else).

I've considered adding a common naming convention for use in `metal-cli`, if we adopted a pattern like this our TF Modules would want to agree:

```
$ metal env --terraform
TF_VAR_metal_auth_token=...
TF_VAR_metal_project_id=...
TF_VAR_metal_organization_id=...

$ metal env --terraform -o tf >> terraform.tfvars
```

cprivitere replied:
What about not using the TF_VAR in front of it, so folks can just export $(metal env) ?

displague replied:
`eval $(metal env)` already exists (no `TF_VAR_` prefix). Terraform environment variables have to follow a particular convention: https://developer.hashicorp.com/terraform/cli/config/environment-variables#tf_var_name

ocobles replied:
https://github.com/equinix-labs/terraform-equinix-kubernetes-addons/blob/8e7ce00af6d2b0379a4ed859c192b82fcc73a4e2/.utils/addon_specific_templates/examples/simple/variables.tf.tpl#L11-L20

I've replaced them with "metal_auth_token" and "metal_project_id" for me it makes sense, and in fact for metal_auth_token it will also be possible to define METAL_AUTH_TOKEN instead of TF_VAR_metal_auth_token 

https://registry.terraform.io/providers/equinix/equinix/latest/docs#:~:text=This%20argument%20can%20also%20be%20specified%20with%20the%20EQUINIX_API_TOKEN%20shell%20environment%20variable.

displague replied:
I think `TF_VAR_metal_auth_token` is the var name

ocobles replied:
Oh, I see I put it wrong in the description. I'll update it

ctreatma commented:
I ran into a couple issues with the script.  I ran:

```bash
% ./generate_addon_structure.sh -n cloud_provider_equinix_metal -w 'https://github.com/equinix/cloud-provider-equinix-metal
```

It stripped the underscores out of `cloud_provider_equinix_metal`, instead creating the `cloudproviderequinixmetal` add-on, and writing the templates failed with an error:

```
sed: 1: "s/{ADDON_WEBSITE_URL}/h ...": bad flag in substitute command: '/'
```

To work around that failure, I set `-w cloud-provider-equinix-metal` and updated the URL manually after generating.

ctreatma commented on .utils/generate_addon_structure.sh:
```diff
@@ -0,0 +1,186 @@
+
+function get_equinix_provider_latest_release() {
+    EQUINIX_PROVIDER_REPO="equinix/terraform-provider-equinix"
+    EQUINIX_PROVIDER_VERSION=$(curl --silent "https://api.github.com/repos/$EQUINIX_PROVIDER_REPO/releases/latest" | grep '"tag_name":' | sed -E 's/.*"([^"]+)".*/\1/')
```
Chris noticed that this is generating an invalid version string because the release tag starts with a `v`: https://github.com/equinix-labs/terraform-equinix-kubernetes-addons/pull/2/commits/edfe9f2e5ed87566a9029e9cee410e01590a6c31

ocobles replied:
Fixed, thanks! 

https://github.com/equinix-labs/terraform-equinix-kubernetes-addons/blob/8e7ce00af6d2b0379a4ed859c192b82fcc73a4e2/.utils/generate_addon_structure.sh#L107

Commit:
update metal variables name to reduce ambiguity

Commit:
fix variable substitutions in templates and use semver equinix provider version

ocobles commented:
> I ran into a couple issues with the script. I ran:
> 
> ```shell
> % ./generate_addon_structure.sh -n cloud_provider_equinix_metal -w 'https://github.com/equinix/cloud-provider-equinix-metal
> ```
> 
> It stripped the underscores out of `cloud_provider_equinix_metal`, instead creating the `cloudproviderequinixmetal` add-on, and writing the templates failed with an error:
> 
> ```
> sed: 1: "s/{ADDON_WEBSITE_URL}/h ...": bad flag in substitute command: '/'
> ```
> 
> To work around that failure, I set `-w cloud-provider-equinix-metal` and updated the URL manually after generating.

@ctreatma 

~Now both " " and "_" will be replaced with "-", so "-n cloud_provider_equinix_metal" will generate a "cloud-provider-equinix-metal" addon.~

EDIT: 
As we have agreed, now for any kind of resources in templates, both " " and "-" will be replaced with "_" , and any other special character will be removed, but the name of the module folder will have "-". 

So using, for example, "**-n cloud provider_equinix%?metal!**" it will generate a "**cloud-provider-equinixmetal**" addon folder and the enable_ variable will be "**enable_cloud_provider_equinixmetal**".

https://github.com/equinix-labs/terraform-equinix-kubernetes-addons/blob/3e32297b5a9f2df2458539b2d504d990fec31ee6/.utils/generate_addon_structure.sh#L173-L174

For the ADDON_WEBSITE_URL issue, I've replaced the sed "/" delimiter with "|"

https://github.com/equinix-labs/terraform-equinix-kubernetes-addons/blob/8e7ce00af6d2b0379a4ed859c192b82fcc73a4e2/.utils/generate_addon_structure.sh#L95

ctreatma commented:
> Now both " " and "_" will be replaced with "-", so "-n cloud_provider_equinix_metal" will generate a "cloud-provider-equinix-metal" addon.

@ocobleseqx Replacing spaces in the addon name sounds good.  Do we have a documented standard/reasoning for replacing `_` with `-`?  I don't think there's a general standard around that, although in practice I've seen teams stick with `_` in variable/resource/datasource names, and only use `-` in variable/attribute values or directory names.

ctreatma commented:
Could you also document the existence & usage of this generator in the README?  I'm assuming it's intended to be run from inside the `.utils` directory (i.e., users should run `% cd .utils; ./generate_addon_structure.sh ...` as opposed to `% .utils/generate_addon_structure.sh ...`)

Commit:
rename all resources in templates to use snake_case and variable addon_config with addon name

Commit:
added contribution instructions

ocobles commented:
> Could you also document the existence & usage of this generator in the README? I'm assuming it's intended to be run from inside the `.utils` directory (i.e., users should run `% cd .utils; ./generate_addon_structure.sh ...` as opposed to `% .utils/generate_addon_structure.sh ...`)

@ctreatma please checkout the instructions added in  

https://github.com/equinix-labs/terraform-equinix-kubernetes-addons/pull/5/commits/cceafb77206e42b7704dc203184a6c00d20e85a0

Commit:
fixup! update metal variables name to reduce ambiguity

cprivitere committed:
remove terraform from module name and convert to snake_case

Signed-off-by: Chris Privitere <23177737+cprivitere@users.noreply.github.com>

Commit:
change delimiters for tf_docs templates

ctreatma commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+# Hello Contributors
+
+Thx for your interest! We're so glad you're here.
```
We could adopt the preamble from @dcallao's [template PR](https://github.com/equinix-labs/terraform-equinix-template/pull/4/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R1-R6) here

ocobles replied:
I'm going to copy the `Contributing` intro and perhaps that `File Structure` section adapted to this project, but I think it makes sense to keep the `Contributing a new Addon` specific to this project.  The `Contributing via Pull Request` sections could also fit in the @dcallao's template. What do you think?

ctreatma commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+
+## Code of Conduct
+
+Available via [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md).
```
I don't see a code of conduct in this repo or PR yet.  Should we pull over the one from @dcallao's [template PR](https://github.com/equinix-labs/terraform-equinix-template/pull/4/files#diff-ffdbe3a1e7ee93cacfc080b6c635ccf3a8f6b0f00f2fb884f78c6b5f9dac8fd2)?

ocobles replied:
Sure, let's put it in a separated PR

ocobles replied:
added in https://github.com/equinix-labs/terraform-equinix-kubernetes-addons/pull/33

cprivitere committed:
Merge branch 'main' into tool_generate_new_addon_layout

github-actions[bot] committed:
terraform-docs: automated action
